presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -2515,42 +2534,274 @@ private void analyzeColumnMask(String currentIdentity, Table table, QualifiedObj
+                                });
+
+                                // c) validate recursion step has no illegal clauses
+                                validateFromClauseOfRecursiveTerm(from, withQuery.getName());
+
+                                // shape validation complete - process query as expandable query
+                                Scope parentScope = withScopeBuilder.build();
+                                // process expandable query -- anchor
+                                Scope anchorScope = process(anchor, parentScope);
+                                // set aliases in anchor scope as defined for WITH query. Recursion step will refer to anchor fields by aliases.
+                                Scope aliasedAnchorScope = setAliases(anchorScope, withQuery.getName(), withQuery.getColumnNames().get());
+                                // record expandable query base scope for recursion step analysis
+                                Node recursiveReference = fromReferences.get(0);
+                                analysis.setExpandableBaseScope(recursiveReference, aliasedAnchorScope);
+                                // process expandable query -- recursion step
+                                Scope stepScope = process(step, parentScope);
+
+                                // verify anchor and step have matching descriptors
+                                RelationType anchorType = aliasedAnchorScope.getRelationType().withOnlyVisibleFields();
+                                RelationType stepType = stepScope.getRelationType().withOnlyVisibleFields();
+                                if (anchorType.getVisibleFieldCount() != stepType.getVisibleFieldCount()) {
+                                    throw semanticException(TYPE_MISMATCH, step, "base and step relations of recursion have different number of fields: %s, %s", anchorType.getVisibleFieldCount(), stepType.getVisibleFieldCount());
+                                }
 
-                String name = withQuery.getName().getValue().toLowerCase(ENGLISH);
-                if (withScopeBuilder.containsNamedQuery(name)) {
-                    throw semanticException(DUPLICATE_NAMED_QUERY, withQuery, "WITH query name '%s' specified more than once", name);
-                }
+                                List<Type> anchorFieldTypes = anchorType.getVisibleFields().stream()
+                                        .map(Field::getType)
+                                        .collect(toImmutableList());
+                                List<Type> stepFieldTypes = stepType.getVisibleFields().stream()
+                                        .map(Field::getType)
+                                        .collect(toImmutableList());
+                                if (!anchorFieldTypes.equals(stepFieldTypes)) {
+                                    analysis.addRelationCoercion(step, anchorFieldTypes.toArray(Type[]::new));
+                                }
```

Comment:
Move this below the block that checks for type compatibility. It doesn't change the outcome, but it's "logically" more correct.